File: backend/school_service/api/routes/teachers.py
"""API routes for Teacher management."""

import asyncio
import json
from typing import Optional, List

from fastapi import APIRouter, Depends, HTTPException, Query, UploadFile, File, status
from fastapi.responses import StreamingResponse
from sqlalchemy.ext.asyncio import AsyncSession

from school_service.core.database import get_db
from school_service.api.routes.auth import get_current_user
from school_service.services.teacher_service import TeacherService
from school_service.schemas.teacher import (
    TeacherCreate,
    TeacherUpdate,
    TeacherResponse,
    TeacherBulkRow,
    TeacherBulkImportRequest,
    TeacherBulkImportResult,
    PaginatedTeacherResponse,
)
from shared.schemas.user import UserResponse

router = APIRouter(prefix="/api/v1/teachers", tags=["teachers"])


# ---------------------------------------------------------------------------
# List
# ---------------------------------------------------------------------------

@router.get(
    "",
    response_model=PaginatedTeacherResponse,
    summary="List teachers",
)
async def list_teachers(
    search: Optional[str] = Query(None, description="Search by name, employee ID, phone, or subject"),
    department: Optional[str] = Query(None),
    is_active: Optional[bool] = Query(None),
    page: int = Query(1, ge=1),
    page_size: int = Query(50, ge=1, le=200),
    current_user: UserResponse = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    svc = TeacherService(db)
    return await svc.list_teachers(
        school_id=current_user.school_id,
        search=search,
        department=department,
        is_active=is_active,
        page=page,
        page_size=page_size,
    )


# ---------------------------------------------------------------------------
# Single teacher CRUD
# ---------------------------------------------------------------------------

@router.get("/{teacher_id}", response_model=TeacherResponse, summary="Get teacher by ID")
async def get_teacher(
    teacher_id: int,
    current_user: UserResponse = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    svc = TeacherService(db)
    try:
        return await svc.get_teacher(teacher_id, current_user.school_id)
    except LookupError as e:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=str(e))


@router.post("", response_model=TeacherResponse, status_code=status.HTTP_201_CREATED, summary="Create teacher")
async def create_teacher(
    data: TeacherCreate,
    current_user: UserResponse = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    svc = TeacherService(db)
    try:
        return await svc.create_teacher(current_user.school_id, data)
    except ValueError as e:
        raise HTTPException(status_code=status.HTTP_409_CONFLICT, detail=str(e))


@router.put("/{teacher_id}", response_model=TeacherResponse, summary="Update teacher")
async def update_teacher(
    teacher_id: int,
    data: TeacherUpdate,
    current_user: UserResponse = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    svc = TeacherService(db)
    try:
        return await svc.update_teacher(teacher_id, current_user.school_id, data)
    except LookupError as e:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=str(e))
    except ValueError as e:
        raise HTTPException(status_code=status.HTTP_409_CONFLICT, detail=str(e))


@router.delete("/{teacher_id}", status_code=status.HTTP_204_NO_CONTENT, summary="Delete teacher")
async def delete_teacher(
    teacher_id: int,
    current_user: UserResponse = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    svc = TeacherService(db)
    try:
        await svc.delete_teacher(teacher_id, current_user.school_id)
    except LookupError as e:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=str(e))


# ---------------------------------------------------------------------------
# Bulk import — JSON rows
# ---------------------------------------------------------------------------

@router.post(
    "/import/json",
    response_model=TeacherBulkImportResult,
    summary="Bulk import teachers from JSON array",
)
async def bulk_import_json(
    body: TeacherBulkImportRequest,
    current_user: UserResponse = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    svc = TeacherService(db)
    return await svc.bulk_import(current_user.school_id, body.teachers)


# ---------------------------------------------------------------------------
# Bulk import — File upload (CSV / TSV)
# ---------------------------------------------------------------------------

@router.post(
    "/import/file",
    summary="Bulk import teachers from CSV, TSV, or XLSX file (streams NDJSON progress)",
)
async def bulk_import_file(
    file: UploadFile = File(..., description="CSV, TSV, or XLSX file"),
    current_user: UserResponse = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    content = await file.read()
    try:
        from school_service.api.file_parser import parse_upload_to_rows
        raw_rows = parse_upload_to_rows(content, file.filename or "")
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))

    rows: List[TeacherBulkRow] = []
    errors: List[str] = []
    for i, normalised in enumerate(raw_rows, start=2):
        try:
            row = TeacherBulkRow(
                first_name=normalised.get("first_name", ""),
                last_name=normalised.get("last_name", ""),
                phone=normalised.get("phone") or normalised.get("phone_number", ""),
                email=normalised.get("email") or None,
                subject=normalised.get("subject") or None,
                department=normalised.get("department") or None,
            )
            rows.append(row)
        except Exception as exc:
            errors.append(f"Row {i}: {exc}")

    total = len(rows)
    queue: asyncio.Queue = asyncio.Queue()

    async def progress_cb(processed: int, total_rows: int) -> None:
        pct = round(100 * processed / total_rows) if total_rows else 0
        await queue.put({"type": "progress", "pct": pct, "processed": processed, "total": total_rows})

    async def run_import() -> None:
        try:
            svc = TeacherService(db)
            result = await svc.bulk_import(
                current_user.school_id, rows, progress_callback=progress_cb
            )
            result.errors = errors + result.errors
            result.skipped += len(errors)
            await queue.put({"type": "result", **result.model_dump()})
        except Exception as e:
            await queue.put({"type": "error", "detail": str(e)})
        await queue.put(None)

    async def stream_ndjson() -> str:
        if total > 0:
            yield json.dumps({"type": "progress", "pct": 0, "processed": 0, "total": total}) + "\n"
        task = asyncio.create_task(run_import())
        try:
            while True:
                msg = await queue.get()
                if msg is None:
                    break
                yield json.dumps(msg) + "\n"
        finally:
            await task

    return StreamingResponse(stream_ndjson(), media_type="application/x-ndjson")

File: frontend/lib/api/teachers.ts
/**
 * API client functions for Teacher management.
 *
 * Base URL: process.env.NEXT_PUBLIC_API_URL || 'http://localhost:8000'
 * Endpoints:
 * - GET    /api/v1/teachers               — List teachers (paginated, search)
 * - GET    /api/v1/teachers/:id           — Get single teacher
 * - POST   /api/v1/teachers               — Create teacher
 * - PUT    /api/v1/teachers/:id           — Update teacher
 * - DELETE /api/v1/teachers/:id           — Soft-delete teacher
 * - POST   /api/v1/teachers/import/json  — Bulk import (JSON)
 * - POST   /api/v1/teachers/import/file  — Bulk import (CSV/TSV)
 */

import axios from "axios";

const API_BASE_URL = process.env.NEXT_PUBLIC_API_URL || "http://localhost:8000";

// -------------------------------------------------------------------------
// Types
// -------------------------------------------------------------------------

export interface TeacherResponse {
    id: number;
    school_id: number;
    employee_id: string | null;
    first_name: string;
    last_name: string;
    phone: string;
    email: string | null;
    subject: string[] | null;
    department: string | null;
    is_active: boolean;
    is_deleted: boolean;
    created_at: string;
    updated_at: string | null;
}

export interface PaginatedTeacherResponse {
    items: TeacherResponse[];
    total: number;
    page: number;
    page_size: number;
    total_pages: number;
}

export interface TeacherCreateData {
    first_name: string;
    last_name: string;
    phone: string;
    email?: string | null;
    subject?: string[] | null;
    department?: string | null;
}

export interface TeacherUpdateData {
    first_name?: string;
    last_name?: string;
    phone?: string;
    email?: string | null;
    subject?: string[] | null;
    department?: string | null;
    is_active?: boolean;
}

export interface ListTeachersParams {
    search?: string;
    department?: string;
    is_active?: boolean;
    page?: number;
    page_size?: number;
}

export interface TeacherBulkImportResult {
    inserted: number;
    updated: number;
    skipped: number;
    errors: string[];
    total: number;
}

// -------------------------------------------------------------------------
// Helpers
// -------------------------------------------------------------------------

function authHeaders(token: string) {
    return {
        "Content-Type": "application/json",
        Authorization: `Bearer ${token}`,
    };
}

// -------------------------------------------------------------------------
// API functions
// -------------------------------------------------------------------------

/** List teachers with optional search/filter. */
export async function listTeachers(
    token: string,
    params: ListTeachersParams = {}
): Promise<PaginatedTeacherResponse> {
    const q = new URLSearchParams();
    if (params.search) q.append("search", params.search);
    if (params.department) q.append("department", params.department);
    if (params.is_active !== undefined) q.append("is_active", String(params.is_active));
    if (params.page) q.append("page", String(params.page));
    if (params.page_size) q.append("page_size", String(params.page_size));

    const { data } = await axios.get<PaginatedTeacherResponse>(
        `${API_BASE_URL}/api/v1/teachers?${q.toString()}`,
        { headers: authHeaders(token) }
    );
    return data;
}

/** Get a single teacher by ID. */
export async function getTeacher(token: string, teacherId: number): Promise<TeacherResponse> {
    const { data } = await axios.get<TeacherResponse>(
        `${API_BASE_URL}/api/v1/teachers/${teacherId}`,
        { headers: authHeaders(token) }
    );
    return data;
}

/** Create a new teacher. */
export async function createTeacher(
    token: string,
    teacherData: TeacherCreateData
): Promise<TeacherResponse> {
    const { data } = await axios.post<TeacherResponse>(
        `${API_BASE_URL}/api/v1/teachers`,
        teacherData,
        { headers: authHeaders(token) }
    );
    return data;
}

/** Update an existing teacher. */
export async function updateTeacher(
    token: string,
    teacherId: number,
    teacherData: TeacherUpdateData
): Promise<TeacherResponse> {
    const { data } = await axios.put<TeacherResponse>(
        `${API_BASE_URL}/api/v1/teachers/${teacherId}`,
        teacherData,
        { headers: authHeaders(token) }
    );
    return data;
}

/** Soft-delete a teacher. */
export async function deleteTeacher(token: string, teacherId: number): Promise<void> {
    await axios.delete(`${API_BASE_URL}/api/v1/teachers/${teacherId}`, {
        headers: authHeaders(token),
    });
}

/** Timeout for bulk file import (10 min) - backend streams NDJSON progress. */
const BULK_IMPORT_TIMEOUT_MS = 600_000;

/**
 * Bulk import teachers from a File object (CSV, TSV).
 * Consumes NDJSON stream from backend for real-time progress updates.
 */
export async function bulkImportTeachersFile(
    token: string,
    file: File,
    onProgress?: (percent: number) => void
): Promise<TeacherBulkImportResult> {
    const formData = new FormData();
    formData.append("file", file);

    const ac = new AbortController();
    const to = setTimeout(() => ac.abort(), BULK_IMPORT_TIMEOUT_MS);

    try {
        const res = await fetch(`${API_BASE_URL}/api/v1/teachers/import/file`, {
            method: "POST",
            headers: { Authorization: `Bearer ${token}` },
            body: formData,
            signal: ac.signal,
        });

        if (!res.ok && res.headers.get("content-type")?.includes("application/json")) {
            const j = await res.json();
            throw new Error(j?.detail ?? "Import failed");
        }
        if (!res.ok) throw new Error(`Import failed: ${res.statusText}`);

        const reader = res.body?.getReader();
        if (!reader) throw new Error("No response body");

        const decoder = new TextDecoder();
        let buffer = "";

        while (true) {
            const { done, value } = await reader.read();
            if (done) break;
            buffer += decoder.decode(value, { stream: true });
            const lines = buffer.split("\n");
            buffer = lines.pop() ?? "";

            for (const line of lines) {
                const trimmed = line.trim();
                if (!trimmed) continue;
                try {
                    const msg = JSON.parse(trimmed) as {
                        type?: string;
                        pct?: number;
                        detail?: string;
                        inserted?: number;
                        updated?: number;
                        skipped?: number;
                        errors?: string[];
                        total?: number;
                    };
                    if (msg.type === "progress" && typeof msg.pct === "number" && onProgress) {
                        onProgress(Math.min(100, Math.max(0, msg.pct)));
                    } else if (msg.type === "result") {
                        return {
                            inserted: msg.inserted ?? 0,
                            updated: msg.updated ?? 0,
                            skipped: msg.skipped ?? 0,
                            errors: msg.errors ?? [],
                            total: msg.total ?? 0,
                        };
                    } else if (msg.type === "error") {
                        throw new Error(msg.detail ?? "Import failed");
                    }
                } catch (e) {
                    // Ignore JSON parse errors for malformed chunks; rethrow our errors
                    if (!(e instanceof SyntaxError)) throw e;
                }
            }
        }

        if (buffer.trim()) {
            const msg = JSON.parse(buffer.trim()) as {
                type?: string;
                detail?: string;
                inserted?: number;
                updated?: number;
                skipped?: number;
                errors?: string[];
                total?: number;
            };
            if (msg.type === "result")
                return {
                    inserted: msg.inserted ?? 0,
                    updated: msg.updated ?? 0,
                    skipped: msg.skipped ?? 0,
                    errors: msg.errors ?? [],
                    total: msg.total ?? 0,
                };
            if (msg.type === "error") throw new Error(msg.detail ?? "Import failed");
        }

        throw new Error("Import completed with no result");
    } catch (err: unknown) {
        if (err instanceof Error) {
            if (err.name === "AbortError") throw new Error("Upload timed out. Try a smaller file.");
            throw err;
        }
        throw new Error("Import failed");
    } finally {
        clearTimeout(to);
    }
}

/**
 * Bulk import from JSON rows (used when frontend parses XLSX/Access and sends JSON).
 */
export async function bulkImportTeachersJson(
    token: string,
    rows: TeacherCreateData[]
): Promise<TeacherBulkImportResult> {
    const { data } = await axios.post<TeacherBulkImportResult>(
        `${API_BASE_URL}/api/v1/teachers/import/json`,
        { teachers: rows },
        { headers: authHeaders(token) }
    );
    return data;
}
